packages/origin-247-certificate/src/evented-certificate/certificate.facade.ts

```diff
@@ -0,0 +1,167 @@
+import { Injectable, Inject } from '@nestjs/common';
+import {
+    CERTIFICATE_COMMAND_REPOSITORY,
+    CertificateCommandRepository
+} from './repositories/CertificateCommand/CertificateCommand.repository';
+import {
+    CERTIFICATE_EVENT_REPOSITORY,
+    CertificateEventRepository
+} from './repositories/CertificateEvent/CertificateEvent.repository';
+import { CertificateService } from '../certificate.service';
+import { QueryBus, EventBus } from '@nestjs/cqrs';
+import { GetAllCertificatesQuery, GetCertificateQuery } from '@energyweb/issuer-api';
+import { CertificateEventType } from './events/Certificate.events';
+
+import {
+    ICertificate,
+    IClaimCommand,
+    IIssueCommand,
+    ITransferCommand,
+    ISuccessResponse,
+    BlockchainActionType,
+    BlockchainAction,
+    IIssuedCertificate,
+    IIssueCommandParams
+} from '../types';
+import {
+    CertificateIssuedEvent,
+    CertificateTransferredEvent,
+    CertificateClaimedEvent
+} from './events/Certificate.events';
+
+const CertificateEventVersion = 1;
+
+type CertificateCommand = IIssueCommand<unknown> | ITransferCommand | IClaimCommand;
+@Injectable()
+export class CertificateFacade<T = null> {
+    constructor(
+        @Inject(CERTIFICATE_COMMAND_REPOSITORY)
+        private readonly certCommandRepo: CertificateCommandRepository,
+        @Inject(CERTIFICATE_EVENT_REPOSITORY)
+        private readonly certEventRepo: CertificateEventRepository,
+        private readonly certService: CertificateService,
+        private readonly queryBus: QueryBus,
+        private readonly eventBus: EventBus
+    ) {}
+
+    public async issue(params: IIssueCommandParams<T>): Promise<IIssuedCertificate<T>> {
+        const command = {
+            ...params,
+            fromTime: Math.round(params.fromTime.getTime() / 1000),
+            toTime: Math.round(params.toTime.getTime() / 1000)
+        } as IIssueCommand<T>;
+
+        const savedCommand = await this.certCommandRepo.create({ payload: command });
+        this.validateCommand(command); // this will probably throw
+
+        const internalCertId = this.generateInternalCertificateId();
+        const event = new CertificateIssuedEvent(internalCertId, CertificateEventVersion, command);
+
+        this.eventBus.publish(event);
```

Comment:
I wonder if we need this event for anything?

Reply:
I think it will be better to do it like that: make repository save the event with generic `save` method, and then repository will publish event, that some new event was added, and it will publish like `NewEventAdded` event with `internalId`

Reply:
I disagree with repository publishing events, this is not repository's responsibility.
I will keep it as is for now.

Reply:
Yes, and no. Maybe it's not "extremely" clean, but at least it's practical.

Then - maybe just make a helper function, that will do these two things: publish and save event?

Reply:
Added propagate method that does just that. Let me know if you have better name for it.